service/grails-app/services/org/olf/licenses/LicenseHousekeepingService.groovy

```diff
@@ -0,0 +1,68 @@
+package org.olf.licenses
+
+import grails.gorm.transactions.Transactional
+import org.olf.licenses.License
+import org.olf.licenses.LicenseAmendment
+import com.k_int.web.toolkit.refdata.RefdataValue
+
+/**
+ * This service works at the module level, it's often called without a tenant context.
+ */
+@Transactional
+public class LicenseHousekeepingService {
+
+  public void triggerHousekeeping() {
+    this.checkUnsetValues();
+  }
+
+  private List<List<String>> batchFetchAmendments(final int platformBatchSize, int platformBatchCount) {
+    List<List<String>> amendments = LicenseAmendment.createCriteria().list ([max: platformBatchSize, offset: platformBatchSize * platformBatchCount]) {
+      order 'id'
+    }
+    return amendments
+  }
+
+  private List<List<String>> batchFetchLicenses(final int platformBatchSize, int platformBatchCount) {
+    List<List<String>> licenses = License.createCriteria().list ([max: platformBatchSize, offset: platformBatchSize * platformBatchCount]) {
+      order 'id'
+    }
+    return licenses
+  }
+
+  def checkUnsetValues() {
+    log.debug("EndDateSemanticsCleanup: Check for unset values")
+
+    def count = 0
+    def batchSize = 25
+
+    License.withNewTransaction {
+      List<License> licenses = batchFetchLicenses(batchSize, count)
+      while (licenses && licenses.size() > 0) {
+        count++
+        licenses.each {License.findAllByEndDateSemanticsIsNull().each { la ->
```

Comment:
At this point we have a List<License> licenses, so we want to do something for each of those Licenses, ie we need:

```
licenses.each {
  l -> l.endDateSemantics = RefdataValue.lookupOrCreate('endDateSemantics', 'Implicit')
  l.save(flush:true, failOnError:true)
}
```